src/last_sign_state.rs

```diff
@@ -0,0 +1,200 @@
+use abscissa::Error;
+use serde_json;
+use std::{
+    fmt::{self, Display},
+    fs::File,
+    io::prelude::*,
+    path::Path,
+};
+use tendermint::{block, chain};
+
+#[derive(Serialize, Deserialize)]
+struct LastSignData {
+    pub height: i64,
+    pub round: i64,
+    pub step: i8,
+    pub block_id: Option<block::Id>,
+}
+
+const EMPTY_DATA: LastSignData = LastSignData {
+    height: 0,
+    round: 0,
+    step: 0,
+    block_id: None,
+};
+
+pub struct LastSignState {
+    data: LastSignData,
+    file: File,
+    _chain_id: chain::Id,
+}
+
+/// Error type
+#[derive(Debug)]
+pub struct LastSignError(Error<LastSignErrorKind>);
+
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Fail)]
+pub enum LastSignErrorKind {
+    #[fail(display = "height regression")]
+    HeightRegression,
+    #[fail(display = "step regression")]
+    StepRegression,
+    #[fail(display = "round regression")]
+    RoundRegression,
+    #[fail(display = "invalid block id")]
+    DoubleSign,
+}
+
+impl From<Error<LastSignErrorKind>> for LastSignError {
+    fn from(other: Error<LastSignErrorKind>) -> Self {
+        LastSignError(other)
+    }
+}
+
+impl Display for LastSignError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl LastSignState {
+    pub fn load_state(path: &Path, chain_id: chain::Id) -> std::io::Result<LastSignState> {
+        if !path.exists() {
+            let mut lst = LastSignState {
+                data: EMPTY_DATA,
+                file: File::create(path)?,
+                _chain_id: chain_id,
+            };
+            lst.sync_to_disk()?;
+            return Ok(lst);
+        }
+        let mut lst = LastSignState {
+            data: EMPTY_DATA,
+            file: File::open(path)?,
+            _chain_id: chain_id,
+        };
+
+        let mut contents = String::new();
+        lst.file.read_to_string(&mut contents)?;
+        lst.data = serde_json::from_str(&contents)?;
+        Ok(lst)
+    }
+
+    pub fn sync_to_disk(&mut self) -> std::io::Result<()> {
+        self.file
+            .write_all(serde_json::to_string(&self.data).unwrap().as_ref())?;
+        self.file.sync_all()?;
```

Comment:
This could be made a bit more atomic by writing the new state to a separate file then replacing/overwriting the other file (e.g. with `fs::rename`)